### PR TITLE
Remove --boxed

### LINKED ISSUE #383: Remove/enable --boxed
There are some tests marked with `--boxed` which don't seem to run in CI, and several appear to be broken. We should get these tests running and maybe remove the `--boxed` option.

### PR DESCRIPTION
**Pull request by Dreamsorcerer**
Fixes #383.

### CODE REVIEW

**Comment by Dreamsorcerer** [comment]
@webknjaz I'm trying to get these tests running again with pytest-forked. Do you have any ideas on the failing tests?

**Comment by webknjaz** [comment]
> E       aiohttp_devtools.exceptions.AiohttpDevException: The port 8000 is already is use

This is clearly a race condition — several tests (probably from different subprocesses/threads) attempt to listen to port `8000`.

To overcome this sort of problem, I usually try to use an ephemeral port `0` so that the kernel would allocate a random unused port consistently. See if you can achieve that.
One other way may be trying to use UNIX-sockets (file or abstract) where you can use unique paths.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -90,7 +90,7 @@ jobs:
       env:
         PYTHONIOENCODING: 'utf-8'
     - name: Run unittests
-      run: pytest
+      run: pytest --forked
```
Comment (webknjaz) [inline .github/workflows/ci.yml:93]: FWIW it's best to put this in `pytest.ini` so that it's shared across invocations (and local executions are equivalent to the CI).

Comment (webknjaz) [reply .github/workflows/ci.yml:93]: I also like to have the config annotated like this: https://github.com/cherrypy/cheroot/blob/e1a45fa/pytest.ini, feel free to copy+adapt.

File: tests/test_runserver_config.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ async def test_create_app_wrong_name(tmpworkdir, loop):
     assert excinfo.value.args[0] == "Module 'app.py' does not define a 'missing' attribute/class"
 
 
-@pytest.mark.boxed
```
Comment (webknjaz) [inline tests/test_runserver_config.py:24]: Note that this marker is still defined in the config.

**Comment by Dreamsorcerer** [comment]
> This is clearly a race condition — several tests (probably from different subprocesses/threads) attempt to listen to port `8000`.

I thought they all ran one at a time.

> To overcome this sort of problem, I usually try to use an ephemeral port `0` so that the kernel would allocate a random unused port consistently.

I'm playing with it, and it seems to work, but I'm then getting `ResourceWarning`s instead, so maybe there is just something not closing the port correctly...

**Comment by Dreamsorcerer** [comment]
> I'm playing with it, and it seems to work, but I'm then getting `ResourceWarning`s instead, so maybe there is just something not closing the port correctly...

OK, that was just because a couple of tests didn't stop the runner.

**Comment by webknjaz** [comment]
> I thought they all ran one at a time.

Well, there may be a leftover from the previous test. Maybe the server isn't shut down cleanly and the kernel doesn't yet know that the port is not occupied anymore. Maybe there's a race condition on the syscall level. Maybe you don't set `SO_REUSABLE` in which case the port cannot be used again immediately.

You'll only understand what's happening exactly, when armed with a debugger.

**Comment by webknjaz** [comment]
> OK, that was just because a couple of tests didn't stop the runner.

Yeah, that happens. Maybe add `await asyncio.sleep(0.25)` before shutting down the loop (this helps with TLS connections in aiohttp — https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown).

**Comment by Dreamsorcerer** [comment]
> > OK, that was just because a couple of tests didn't stop the runner.
> 
> Yeah, that happens. Maybe add `await asyncio.sleep(0.25)` before shutting down the loop (this helps with TLS connections in aiohttp — https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown).

I mean, someone just forgot to do `await runner.cleanup()`. Appears to be working now.

File: Makefile [hunk 3]
```diff
@@ -15,15 +15,11 @@ lint:
 
 .PHONY: test
 test:
-	pytest --cov=aiohttp_devtools --boxed --duration 5 && (coverage combine || test 0)
-
-.PHONY: testcov
-testfast:
-	pytest --cov=aiohttp_devtools --boxed --fast -n 4 && (echo "building coverage html"; coverage combine; coverage html)
+	pytest --cov=aiohttp_devtools --forked --duration 5 && (coverage combine || test 0)
```
Comment (webknjaz) [inline Makefile:18]: Why duplicate this in every script when you can just put it into the config?

File: requirements-dev.txt [hunk 4]
```diff
@@ -8,10 +8,10 @@ mypy==0.930
 pytest==6.2.5
 pytest-aiohttp==0.3.0
 pytest-cov==3.0.0
+pytest-forked==1.4.0
```
Comment (webknjaz) [inline requirements-dev.txt:11]: FYI: https://github.com/pytest-dev/pytest-forked/issues/44

File: requirements-dev.txt [hunk 5]
```diff
@@ -8,10 +8,10 @@ mypy==0.930
...
 pytest-mock==3.6.1
 pytest-sugar==0.9.4
 pytest-timeout==2.0.2
 pytest-toolbox==0.4
-pytest-xdist==2.5.0
```
Comment (webknjaz) [inline requirements-dev.txt:15]: Why don't you like xdist, by the way? Any tests using signals?

File: Makefile (hunk 3 above)
Comment (Dreamsorcerer) [reply Makefile:18]: I don't know. I'm happy to just remove the whole thing. I'm not using it and it doesn't appear to be used anywhere else.

File: requirements-dev.txt (hunk 5 above)
Comment (Dreamsorcerer) [reply requirements-dev.txt:15]: I was under the impression it was only used for `--boxed` and the documentation suggested this functionality had been moved to pytest-forked. I don't really know anything about it.

Comment (webknjaz) [reply requirements-dev.txt:15]: Right. I don't think it ever worked taking into account that it's also used in windows. FWIW `xdist` creates several "workers" and each worker executes a number of tests subsequently while `forked` creates subprocesses for every or just marked tests.

File: Makefile (hunk 3 above)
Comment (webknjaz) [reply Makefile]: Note that `--forked` / `--boxed` essentially isolates every test but the marker decorators `@pytest.mark.forked` only do that to specific tests: https://github.com/pytest-dev/pytest-forked#usage-examples.

It looks like the CLI arg is just unnecessary.

File: Makefile [hunk 6]
```diff
@@ -4,7 +4,7 @@
 install:
 	pip install -U setuptools pip
 	pip install -e .
-	pip install -r tests/requirements.txt
+	pip install -r requirements-dev.txt
```
Comment (webknjaz) [inline Makefile:7]: Why did you move this? Is it used in things other than tests?

File: aiohttp_devtools/logs.py [hunk 7]
```diff
@@ -44,7 +43,7 @@ def __init__(self, fmt: Optional[str] = None, datefmt: Optional[str] = None, sty
         super().__init__(fmt, datefmt, style)
         self.stream_is_tty = False
 
-    def format(self, record: LogRecord) -> str:
+    def format(self, record: logging.LogRecord) -> str:
```
Comment (webknjaz) [inline aiohttp_devtools/logs.py:46]: This should be separate.

File: aiohttp_devtools/runserver/config.py [hunk 8]
```diff
@@ -154,8 +154,8 @@ def import_app_factory(self) -> AppFactory:
             raise AdevConfigError("'{}.{}' is not an Application or callable".format(
                 self.py_file.name, self.app_factory_name))
 
-        if callable(attr):
-            required_args = attr.__code__.co_argcount - len(attr.__defaults__)
+        if hasattr(attr, "__code__"):
+            required_args = attr.__code__.co_argcount - len(attr.__defaults__ or ())
```
Comment (webknjaz) [inline aiohttp_devtools/runserver/config.py:158]: Separate?

File: aiohttp_devtools/runserver/main.py [hunk 9]
```diff
@@ -28,8 +28,7 @@ class RunServer(TypedDict):
 
 
 def runserver(**config_kwargs: Any) -> RunServer:
-    """
-    Prepare app ready to run development server.
+    """Prepare app ready to run development server.
```
Comment (webknjaz) [inline aiohttp_devtools/runserver/main.py:31]: Seems like there's a ton of unrelated refactoring which doesn't exactly help review things.

File: setup.cfg [hunk 10]
```diff
@@ -1,10 +1,8 @@
 [tool:pytest]
-addopts = --cov=aiohttp_devtools --cov-report xml --cov-report html --cov-report term
-markers = boxed: mark tests as boxed
+addopts = --cov=aiohttp_devtools --cov-report xml --cov-report html --cov-report term --forked
```
Comment (webknjaz) [inline setup.cfg:2]:
```suggestion
addopts = --cov=aiohttp_devtools --cov-report xml --cov-report html --cov-report term
```
(so that only marked tests are isolated)

File: tests/test_runserver_config.py [hunk 11]
```diff
@@ -52,3 +52,16 @@ def app_factory():
...
+@forked
+async def test_wrong_function_signature(tmpworkdir):
+    mktree(tmpworkdir, {
+        'app.py': """\
+def app_factory(foo):
+    return web.Application()
+"""
+    })
+    config = Config(app_path='app.py')
+    with pytest.raises(AiohttpDevConfigError):
```
Comment (webknjaz) [inline tests/test_runserver_config.py:66]: Always use the `, match=` arg with `pytest.raises()` and `pytest.warns()` CMs.

File: requirements-dev.txt (hunk 5 above)
Comment (webknjaz) [reply requirements-dev.txt:15]: FWIW try keeping the parallelism. Just add `--numprocesses=auto` to the config.

File: aiohttp_devtools/runserver/main.py (hunk 9 above)
Comment (Dreamsorcerer) [reply aiohttp_devtools/runserver/main.py:31]: Yep, sorry. Several changes are also to fix issues highlighted by the forked tests which haven't been running for the past couple of years.

File: setup.cfg (hunk 10 above)
Comment (Dreamsorcerer) [reply setup.cfg]: When I do this, it doesn't seem to fork any of the tests. Is there something else I'm meant to add to the config?

File: requirements-dev.txt (hunk 5 above)
Comment (webknjaz) [reply requirements-dev.txt:15]: > I was under the impression it was only used for `--boxed`

It was also used for `-n4` in `Makefile` but you removed those entries. This is why I suggested putting this setting into the config with `auto`.

File: aiohttp_devtools/runserver/main.py (hunk 9 above)
Comment (webknjaz) [reply aiohttp_devtools/runserver/main.py:31]: Oh, I guess I'd still want them separate for atomicity.

File: setup.cfg (hunk 10 above)
Comment (webknjaz) [reply setup.cfg]: No, it should just work. But I'm not sure if it'll be visible in the report at all.

Comment (Dreamsorcerer) [reply setup.cfg]: Well, the CI is failing as it does on my machine, which I believe is because the tests are not forking.

Comment (webknjaz) [reply setup.cfg]: I've just verified that pytest-forked tests this case: https://github.com/pytest-dev/pytest-forked/blob/f10fb75/testing/test_boxed.py#L22-L34. So I'm pretty sure it should work.

Comment (webknjaz) [reply setup.cfg]: Maybe it's caused by other tests not being marked? (since `--forked` would've marked _every_ test regardless of the decorator)

Comment (Dreamsorcerer) [reply setup.cfg]: Yep, that was it. Curious how they ever worked in the past...

File: Makefile (hunk 6 above)
Comment (Dreamsorcerer) [reply Makefile:7]: Includes mypy etc. too. Just making it more consistent with other repos.

File: requirements-dev.txt (hunk 5 above)
Comment (Dreamsorcerer) [reply requirements-dev.txt:15]: The tests take 3-5 seconds to run. It's not worth thinking about.

File: Makefile (hunk 6 above)
Comment (webknjaz) [reply Makefile]: I guess that's okay for now, then. Maybe someday this project will be ready to make use of more precise per-env lockfiles (similar to what I'm trying to achieve with https://github.com/cherrypy/cheroot/tree/e1a45fa/requirements).
